include/dali/dali.h

```diff
@@ -0,0 +1,698 @@
+  const daliDeleter_t *sample_deleters);
+
+
+/** Associates a stream with the TensorList.
+ *
+ * @param tensor_list   [in]  the TensorList whose buffer placement is queried
+ * @param out_placement [out] a pointer to a place where the return value is stored.
+ */
+DALI_API daliResult_t daliTensorListGetBufferPlacement(
+  daliTensorList_h tensor_list,
+  daliBufferPlacement_t *out_placement);
+
+/** Associates a stream with the TensorList.
+ *
+ * @param stream      an optional CUDA stream handle; if the handle poitner is NULL,
+ *                    host-synchronous behavior is prescribed.
+ * @param synchronize if true, the new stream (or host, if NULL), will be synchronized with the
+ *                    currently associated stream
+ */
+DALI_API daliResult_t daliTensorListSetStream(
+  daliTensorList_h tensor_list,
+  const cudaStream_t *stream,
+  daliBool synchronize
+);
+
+/** Gets the stream associated with the TensorList.
+ *
+ * @retval DALI_SUCCESS if the stream handle was stored in *out_stream
+ * @retval DALI_NO_DATA if the tensor list is not associated with any stream
+ *         error code otherwise
+ */
+DALI_API daliResult_t daliTensorListGetStream(
+  daliTensorList_h tensor_list,
+  cudaStream_t *out_stream
+);
+
+/** Gets the readiness event associated with the TensorList.
+ *
+ * @param tensor_list [in]  the tenosr list whose ready event is to be obtained
+ * @param out_event   [out] the pointer to the return value
+ *
+ * @retval DALI_SUCCESS if the ready event handle was stored in *out_event
+ * @retval DALI_NO_DATA if the tensor list is not associated with a readiness event
+ *         error code otherwise
+ */
+DALI_API daliResult_t daliTensorListGetReadyEvent(
+  daliTensorList_h tensor_list,
+  cudaEvent_t *out_event);
+
+/** Gets the readiness event associated with the TensorList or creates a new one.
+ *
+ * @param tensor_list   [in]  the tensor list to associate an even twith
+ * @param out_event     [out] optional, the event handle
+ *
+ * The function ensures that a readiness event is associated with the tensor list.
+ * It can also get the event handle, if the output parameter pointer is not NULL.
+ */
+DALI_API daliResult_t daliTensorListGetOrCreateReadyEvent(
+  daliTensorList_h tensor_list,
+  cudaEvent_t *out_event);
+
+
+/** Gets the shape of the tensor list
+ *
+ * @param tensor_list     [in]  the tensor list whose shape obtain
+ * @param out_num_samples [out] optional; the number of samples in the batch
+ * @param out_ndim        [out] optional; the number of dimensions in a sample
+ * @param out_shape       [out] optional; the pointer to the concatenated array of sample shapes;
+ *                              contains (*out_num_samples) * (*out_ndim) elements
+ *
+ * @retval DALI_SUCCESS T
+ *
+ * The pointer returned in `out_shape` remains valid until the TensorList is destroyed or modified.
+ * If the caller is not intersted in some of the values, the pointers can be NULL.
```

Comment:
Is it expected that for the call: `daliResult_t retval = daliTensorListGetShape(tl, NULL, NULL, NULL) ;`, the `retval` will be DALI_SUCCESS?

Reply:
Either that or `DALI_NO_DATA`. If we make the fields optional, there's no point in adding an artificial failure in a pointless call.